src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/impl/InsertOperationPage.java

```diff
@@ -15,59 +19,118 @@
-      failedIds.forEach((documentId, throwable) -> errors.add(getError(documentId, throwable)));
+    // Ensure insertions and errors are in the input order (wrt unordered insertions),
+    // regardless of output format
+    Collections.sort(successfulInsertions);
+    if (!failedInsertions().isEmpty()) {
+      Collections.sort(
+          failedInsertions, Comparator.comparing(tuple -> tuple.getItem1().position()));
+    }
 
+    if (!returnDocumentResponses()) { // legacy output, limited to ids, error messages
+      List<CommandResult.Error> errors;
+      if (failedInsertions().isEmpty()) {
+        errors = null;
+      } else {
+        errors =
+            failedInsertions.stream()
+                .map(tuple -> getError(tuple.getItem1().document().id(), tuple.getItem2()))
+                .toList();
+      }
+      // Old style, simple ids:
+      List<DocumentId> insertedIds =
+          successfulInsertions.stream().map(docAndPos -> docAndPos.document().id()).toList();
       return new CommandResult(null, Map.of(CommandStatus.INSERTED_IDS, insertedIds), errors);
     }
 
-    // id no errors, just inserted ids
-    return new CommandResult(Map.of(CommandStatus.INSERTED_IDS, insertedIds));
+    // New style output: detailed responses.
+    InsertionResult[] results = new InsertionResult[allAttemptedInsertions().size()];
+    List<CommandResult.Error> errors = new ArrayList<>();
+
+    // Results array filled in order: first successful insertions
+    for (InsertOperation.WritableDocAndPosition docAndPos : successfulInsertions) {
+      results[docAndPos.position()] =
+          new InsertionResult(docAndPos.document().id(), InsertionStatus.OK, null);
+    }
+    // Second: failed insertions
+    for (Tuple2<InsertOperation.WritableDocAndPosition, Throwable> failed : failedInsertions) {
+      InsertOperation.WritableDocAndPosition docAndPos = failed.getItem1();
+      Throwable throwable = failed.getItem2();
+      CommandResult.Error error = getError(throwable);
+      int errorIdx = errors.indexOf(error);
+      if (errorIdx < 0) { // not yet added, add
+        errorIdx = errors.size();
+        errors.add(error);
+      }
+      int idx = docAndPos.position();
+      results[idx] =
+          new InsertionResult(docAndPos.document().id(), InsertionStatus.ERROR, errorIdx);
+    }
+    // And third, if any, skipped
+    for (int i = 0; i < results.length; i++) {
+      if (null == results[i]) {
+        results[i] =
+            new InsertionResult(
+                allAttemptedInsertions.get(i).document().id(), InsertionStatus.SKIPPED, null);
+      }
+    }
+    return new CommandResult(
+        null, Map.of(CommandStatus.DOCUMENT_RESPONSES, Arrays.asList(results)), errors);
   }
 
   private static CommandResult.Error getError(DocumentId documentId, Throwable throwable) {
     String message =
         "Failed to insert document with _id %s: %s".formatted(documentId, throwable.getMessage());
-
```

Comment:
This code was not being used, not sure why it was there (ThrowableToErrorMapper handles this logic)